### PR TITLE
Added support for LaCrosse LTV-WR1 Multi Sensor

### PR DESCRIPTION
**Pull request by mbruski**

### CODE REVIEW

**Comment by mbruski** [comment]
Let's see if I got this one correct before I add the LTV-TH3.

Note that this decoder outputs good wind speed/direction data but rainfall data is output 'raw_' for now until I discover the magic equation for the raw variables.

File: src/devices/lacrosse_wr1.c [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    // base and/or scale adjustments
+    speed_kmh = raw_wind * 0.1f;
+    rain_mm   = 0.0;  // dummy until we know what raw_rain1 and raw_rain2 mean
+
+    /* clang-format off */
+    data = data_make(
+            "model",            "",                 DATA_STRING, "LaCrosse-WR1",
+            "id",               "Sensor ID",        DATA_FORMAT, "%06x", DATA_INT, id,
+            "seq",              "Sequence",         DATA_FORMAT, "%01x", DATA_INT, seq,
```
Comment (zuckschwerdt) [inline src/devices/lacrosse_wr1.c:121]: `seq` is 3 bit (0-7) a format isn't needed.

**Comment by zuckschwerdt** [comment]
This looks really good.

File: src/devices/lacrosse_wr1.c (hunk 1 above)
Comment (mbruski) [reply src/devices/lacrosse_wr1.c:121]: It has been displaying correctly during testing but if you want to change it, be my guest.

**Comment by mbruski** [comment]
Do you want me to commit additions for LTV-TH3 to this branch or should I create a separate one?

**Comment by some-random-kahuna** [comment]
I have an older Lacrosse sensor, the LTV-TH2, and since my Lacrosse C84343 base station lists both the LTV-TH2 and LTV-TH3 as compatible sensors, I took at look at this recent addition and found that it works with my LTV-TH2 with a couple of changes.  First, the message length appears to be 285 bytes; greater than the 256 byte upper limit check.  Secondly, the CRC transmitted by the sensor has a remainder of zero only with sequence numbers of 2 and 6.  For sequence numbers of 0,1,3,4,5, and 7 the CRC is not 0, but 0xac.  Maybe the message format is slightly different in the TH2?  For my environment , the decoding of messages with those two changes in lacrosse_th3.c is 100% and the data returned corresponds with the temp/humidity shown on the LCD on the sensor.  This works for my purposes, but I'm willing to help with a less dirty hack if needed...Aloha

**Comment by mbruski** [comment]
Thanks for that input.  I can put out a decoder specific for the th2 but it would be helpful to have a few sample files to work with.MikeSent from Samsung Galaxy smartphone.
-------- Original message --------From: some-random-kahuna <notifications@github.com> Date: 11/2/20  20:33  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
I have an older Lacrosse sensor, the LTV-TH2, and since my Lacrosse C84343 base station lists both the LTV-TH2 and LTV-TH3 as compatible sensors, I took at look at this recent addition and found that it works with my LTV-TH2 with a couple of changes.  First, the message length appears to be 285 bytes; greater than the 256 byte upper limit check.  Secondly, the CRC transmitted by the sensor has a remainder of zero only with sequence numbers of 2 and 6.  For sequence numbers of 0,1,3,4,5, and 7 the CRC is not 0, but 0xac.  Maybe the message format is slightly different in the TH2?  For my environment , the decoding of messages with those two changes in lacrosse_th3.c is 100% and the data returned corresponds with the temp/humidity shown on the LCD on the sensor.  This works for my purposes, but I'm willing to help with a less dirty hack if needed...Aloha

**Comment by some-random-kahuna** [comment]
Hi Mike,

Here are two.  The g001 file is a capture that has a sequence of 5 and a CRC of 0xac.  The g002 file is a capture that has a sequence of 6 and a CRC of 0x00:

$ rtl_433 -s 1000k -X 'n=th2, m=FSK_PCM, s=104, l=104, r=9600' g001_915M_1000k.cu8 
rtl_433 version 20.02-212-ga2de0c9 branch master at 202010270814 inputs file rtl_tcp RTL-SDR SoapySDR
Use -h for usage help and see https://triq.org/ for documentation.
Trying conf file at "rtl_433.conf"...
Trying conf file at "/home/pi/.config/rtl_433/rtl_433.conf"...
Trying conf file at "/usr/local/etc/rtl_433/rtl_433.conf"...
Trying conf file at "/etc/rtl_433/rtl_433.conf"...
Registered 144 out of 173 device decoding protocols [ 1-4 8 11-12 15-17 19-21 23 25-26 29-36 38-60 63 67-71 73-100 102-105 108-116 119 121 124-128 130-149 151-161 163-168 170-173 ]
Test mode active. Reading samples from file: g001_915M_1000k.cu8
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : @0.073940s
model     : th2          count     : 1             num_rows  : 1
rows      : 
len       : 286          data      : 555555555555555a5545ba865caae1452c09b27a5a5a5a5a400000000000000000000000
codes     : {286}555555555555555a5545ba865caae1452c09b27a5a5a5a5a400000000000000000000000
message: 0x32 0xe5 0x57 0xa 0x29 0x60 0x4d 0x93 0x58 0x68 0x5f crc: 0xac
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : @0.073940s
model     : LaCrosse-TH3 Sensor ID : 32e557
Sequence  : 5            unknown   : 0             Temperature: 26.2 C
Humidity  : 77 %         Integrity : CRC
++++++++++++++++++++++++++++++++++++++++++++++
 $ rtl_433 -s 1000k -X 'n=th2, m=FSK_PCM, s=104, l=104, r=9600' g002_915M_1000k.cu8 
rtl_433 version 20.02-212-ga2de0c9 branch master at 202010270814 inputs file rtl_tcp RTL-SDR SoapySDR
Use -h for usage help and see https://triq.org/ for documentation.
Trying conf file at "rtl_433.conf"...
Trying conf file at "/home/pi/.config/rtl_433/rtl_433.conf"...
Trying conf file at "/usr/local/etc/rtl_433/rtl_433.conf"...
Trying conf file at "/etc/rtl_433/rtl_433.conf"...
Registered 144 out of 173 device decoding protocols [ 1-4 8 11-12 15-17 19-21 23 25-26 29-36 38-60 63 67-71 73-100 102-105 108-116 119 121 124-128 130-149 151-161 163-168 170-173 ]
Test mode active. Reading samples from file: g002_915M_1000k.cu8
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : @0.073945s
model     : th2          count     : 1             num_rows  : 1
rows      : 
len       : 286          data      : 555555555555555a5545ba865caae1852c09acba5a5a5a5a400000000000000000000000
codes     : {286}555555555555555a5545ba865caae1852c09acba5a5a5a5a400000000000000000000000
message: 0x32 0xe5 0x57 0xc 0x29 0x60 0x4d 0x65 0x58 0xc8 0xca crc:  0
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : @0.073945s
model     : LaCrosse-TH3 Sensor ID : 32e557
Sequence  : 6            unknown   : 0             Temperature: 26.2 C
Humidity  : 77 %         Integrity : CRC

[th2_samples.zip](https://github.com/merbanan/rtl_433/files/5483350/th2_samples.zip)

**Comment by mbruski** [comment]
I just opened the samples.  It looks like the bit length is 107us, not 104us like the th3 uses.   That might explain the difference with the crc.  Let me work on decoder and see if thats all that is different.Sent from Samsung Galaxy smartphone.
-------- Original message --------From: some-random-kahuna <notifications@github.com> Date: 11/3/20  13:42  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
Hi Mike,
Here are two.  The g001 file is a capture that has a sequence of 5 and a CRC of 0xac.  The g002 file is a capture that has a sequence of 6 and a CRC of 0x00:
$ rtl_433 -s 1000k -X 'n=th2, m=FSK_PCM, s=104, l=104, r=9600' g001_915M_1000k.cu8
rtl_433 version 20.02-212-ga2de0c9 branch master at 202010270814 inputs file rtl_tcp RTL-SDR SoapySDR
Use -h for usage help and see https://triq.org/ for documentation.
Trying conf file at "rtl_433.conf"...
Trying conf file at "/home/pi/.config/rtl_433/rtl_433.conf"...
Trying conf file at "/usr/local/etc/rtl_433/rtl_433.conf"...
Trying conf file at "/etc/rtl_433/rtl_433.conf"...
Registered 144 out of 173 device decoding protocols [ 1-4 8 11-12 15-17 19-21 23 25-26 29-36 38-60 63 67-71 73-100 102-105 108-116 119 121 124-128 130-149 151-161 163-168 170-173 ]
Test mode active. Reading samples from file: g001_915M_1000k.cu8

time      : @0.073940s
model     : th2          count     : 1             num_rows  : 1
rows      :
len       : 286          data      : 555555555555555a5545ba865caae1452c09b27a5a5a5a5a400000000000000000000000
codes     : {286}555555555555555a5545ba865caae1452c09b27a5a5a5a5a400000000000000000000000
message: 0x32 0xe5 0x57 0xa 0x29 0x60 0x4d 0x93 0x58 0x68 0x5f crc: 0xac

time      : @0.073940s
model     : LaCrosse-TH3 Sensor ID : 32e557
Sequence  : 5            unknown   : 0             Temperature: 26.2 C
Humidity  : 77 %         Integrity : CRC
++++++++++++++++++++++++++++++++++++++++++++++
$ rtl_433 -s 1000k -X 'n=th2, m=FSK_PCM, s=104, l=104, r=9600' g002_915M_1000k.cu8
rtl_433 version 20.02-212-ga2de0c9 branch master at 202010270814 inputs file rtl_tcp RTL-SDR SoapySDR
Use -h for usage help and see https://triq.org/ for documentation.
Trying conf file at "rtl_433.conf"...
Trying conf file at "/home/pi/.config/rtl_433/rtl_433.conf"...
Trying conf file at "/usr/local/etc/rtl_433/rtl_433.conf"...
Trying conf file at "/etc/rtl_433/rtl_433.conf"...
Registered 144 out of 173 device decoding protocols [ 1-4 8 11-12 15-17 19-21 23 25-26 29-36 38-60 63 67-71 73-100 102-105 108-116 119 121 124-128 130-149 151-161 163-168 170-173 ]
Test mode active. Reading samples from file: g002_915M_1000k.cu8

time      : @0.073945s
model     : th2          count     : 1             num_rows  : 1
rows      :
len       : 286          data      : 555555555555555a5545ba865caae1852c09acba5a5a5a5a400000000000000000000000
codes     : {286}555555555555555a5545ba865caae1852c09acba5a5a5a5a400000000000000000000000
message: 0x32 0xe5 0x57 0xc 0x29 0x60 0x4d 0x65 0x58 0xc8 0xca crc:  0

time      : @0.073945s
model     : LaCrosse-TH3 Sensor ID : 32e557
Sequence  : 6            unknown   : 0             Temperature: 26.2 C
Humidity  : 77 %         Integrity : CRC
th2_samples.zip

**Comment by zuckschwerdt** [comment]
If there is a sufficient preamble (like the `55555555555555`s here) the slicer will automatically pick that up. You could use 80 µs or 150 µs and it will still work.

**Comment by mbruski** [comment]
That's good to know.   Thanks.   I cloned the TH3 and made some changes for the CRC check.  The resulting th2 source is attached if he wants to build it and try it out before I commit it.

Mike

**Comment by some-random-kahuna** [comment]
Happy to check it on my HW, but I'm not seeing the attachment.

**Comment by mbruski** [comment]
lacrosse_th2.c (5.9KB) is attached.

**Comment by mbruski** [comment]
Please send me a PM at michael.bruski@gmail.com.  Github is not allowing me to attach a .c file to emails.

**Comment by zuckschwerdt** [comment]
If it's just changes for the CRC you might consider checking both in the TH3 code an then toggle the `model`.

**Comment by mbruski** [comment]
According to Mark, the CRC only differs for packet sequences 2 & 6.  I only had a seq 5 & 6 to work with so can only confirm for those two.  That said, the change i came up with only has CRC to use as a discriminater because packet length for both devices is the same.  So I can only reliably predict model on seq #2 & #6.  Now, if the slicer would expose the real bit width, that would give me a better way to determine model (in this case).  Unfortunately, I don't see a way to check that.MikeSent from Samsung Galaxy smartphone.
-------- Original message --------From: "Christian W. Zuckschwerdt" <notifications@github.com> Date: 11/4/20  01:25  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
If it's just changes for the CRC you might consider checking both in the TH3 code an then toggle the model.

**Comment by mbruski** [comment]
Sorry, I was up late watching election results and suffering brain fog this morning because of it.I can determine model reliably for all sequences EXCEPT #2 & #6.  My thinking is that a separate decoder for these two is probably the best option.  A couple weeks ago I was thinking its odd that there is no model info embedded in the packet.  What would happen if my neighbor bought a different model from LaCrosse that my console heard and thought it recognized?  How would you pair the console to sensor?  So I was wondering if LaCrosse didn't actually take that into consideration and somehow encode model info into that barcoded device ID?  It would take a pool of ID numbers for various models to figure that out... not sure its worth the time but it might need to be done.  Since I wrote that decoder for the WSDTH01 a couple months ago, LaCrosse has added three more models to their web page; WSDTH02, WSDTH03 & WSDTH04.  The LTV line of products is quickly expanding which might make it increasingly difficult to descriminate some models.MikeSent from Samsung Galaxy smartphone.
-------- Original message --------From: aj9x <aj9x@comcast.net> Date: 11/4/20  07:52  (GMT-05:00) To: merbanan/rtl_433 <reply@reply.github.com>, merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
According to Mark, the CRC only differs for packet sequences 2 & 6.  I only had a seq 5 & 6 to work with so can only confirm for those two.  That said, the change i came up with only has CRC to use as a discriminater because packet length for both devices is the same.  So I can only reliably predict model on seq #2 & #6.  Now, if the slicer would expose the real bit width, that would give me a better way to determine model (in this case).  Unfortunately, I don't see a way to check that.MikeSent from Samsung Galaxy smartphone.-------- Original message --------From: "Christian W. Zuckschwerdt" <notifications@github.com> Date: 11/4/20  01:25  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
If it's just changes for the CRC you might consider checking both in the TH3 code an then toggle the model.

**Comment by zuckschwerdt** [comment]
Timing likely isn't a factor here. The units (receivers and sender) won't have a stable enough reference and thus just lock onto the preamble. That's suggested best practice (not that any designer there ever seems to care about best practice…)

If only the CRC is the difference (and maybe some model bits we don't known about) and even the CRC can overlap (2,6 cases) then it's best to handle it in a single decoder -- having false positives is worse. Maybe even regard the TH2 as a TH3 if there isn't enough evidence to split them: model names refer to a protocol, user will known or not care what specific revision their unit is.

**Comment by some-random-kahuna** [comment]
Maybe that's why there's a TH3: they fixed the firmware that generated a bad CRC 25% of the time in the TH2

**Comment by mbruski** [comment]
You could be right.  They obviously have different people working on these designs.  All use the same transmitter chip that I can tell yet each uses different preamble, packet endings, pulse widths and now CRC.Sent from Samsung Galaxy smartphone.
-------- Original message --------From: some-random-kahuna <notifications@github.com> Date: 11/4/20  20:10  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
Maybe that's why there's a TH3: they fixed the firmware that generated a bad CRC 25% of the time in the TH2

**Comment by some-random-kahuna** [comment]
Hi Mike,

If you're still interested in this, I've been running your TH2 decoder for a while and have noticed at least one malformed message that made it though.  The file it's going into has been awked and I don't have any other information.  Would you like a few more captures to fine tune the decoder timing?

2020-11-20 03:05:16 26.700 75
2020-11-20 04:05:16 226.400 302
2020-11-20 05:05:16 26.500 75

Aloha,
Mark

**Comment by mbruski** [comment]
Hey Mark,I'm kind of puzzled as to why that happened.  Being the values are clearly weird, I wonder if this a special message being sent by the sensor.  I don't think it was corrupt or a timing issue as that should have caused the CRC check to fail.  CRC failure should cause the packet to be dropped.  Someone went into one of LTV decoders a week or two ago and put some limit checks on the values pulled for that sensor.  It might have been the WSDTH.  I can go back and do the same for the TH2/TH3 if you think that solution is acceptable.MikeSent from Samsung Galaxy smartphone.
-------- Original message --------From: some-random-kahuna <notifications@github.com> Date: 11/20/20  16:49  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
Hi Mike,
If you're still interested in this, I've been running your TH2 decoder for a while and have noticed at least one malformed message that made it though.  The file it's going into has been awked and I don't have any other information.  Would you like a few more captures to fine tune the decoder timing?
2020-11-20 03:05:16 26.700 75
2020-11-20 04:05:16 226.400 302
2020-11-20 05:05:16 26.500 75
Aloha,
Mark

**Comment by zuckschwerdt** [comment]
Using the values we should be able to convert them back to the original message, maybe something obvious pops out then?
(e.g. on some decoders we assumed a value to be 8 bit but then the top bit was something like a battery flag).

**Comment by mbruski** [comment]
Good Morning Christian and Mark,

I did as Christian suggested and looked at the bit patterns.  When I did Marks numbers, nothing jumped out at me at first.   Then I went back and guessed that temp should have been 26.6C and humidity 75% and constructed a bit string for that pair.  Comparing the two strings, it looks like 2 bits between the sync word and temp were dropped causing the values I 'expected' to be shifted left by 2.  Could this be because of differences in bit timing (104us for TH3 and 107us for TH2)?  Could it be only certain bit sequences cause problems when the bitbuffer is loaded?   All the same, I would have expected this packet to be tossed because CRC didn't check.

Mike

**Comment by some-random-kahuna** [comment]
Hi Mike,

I've had it running for some time now without any more spurious outputs.  Given that whatever is going on looks like a rare occurance, I think your suggestion of just sanity checking the temp and humidity values makes sense.  Here's what I added:

    // base and/or scale adjustments
    temp_c = (raw_temp - 400) * 0.1f;

    if (temp_c < -40.0 || temp_c > 60.0 || humidity < 10 || humidity > 99) {
        if (decoder->verbose) {
           fprintf(stderr, "%s: Spurious temp and/or humidity values!\n", __func
__);
        }
        return DECODE_FAIL_MIC;
    }

**Comment by mbruski** [comment]
Mark,That's what I had in mind.   The alternative would be to try a dedicated decoder for the TH2 specifying the proper bit timing but this is a simpler solution.  Let's see what Christian Z. thinks.MikeSent from Samsung Galaxy smartphone.
-------- Original message --------From: some-random-kahuna <notifications@github.com> Date: 11/23/20  19:36  (GMT-05:00) To: merbanan/rtl_433 <rtl_433@noreply.github.com> Cc: Mike <aj9x@comcast.net>, Author <author@noreply.github.com> Subject: Re: [merbanan/rtl_433] Added support for LaCrosse LTV-WR1 Multi Sensor (#1533) 
Hi Mike,
I've had it running for some time now without any more spurious outputs.  Given that whatever is going on looks like a rare occurance, I think your suggestion of just sanity checking the temp and humidity values makes sense.  Here's what I added:
// base and/or scale adjustments
temp_c = (raw_temp - 400) * 0.1f;

if (temp_c < -40.0 || temp_c > 60.0 || humidity < 10 || humidity > 99) {
    if (decoder->verbose) {
       fprintf(stderr, "%s: Spurious temp and/or humidity values!\n", __func

__);
}
return DECODE_FAIL_MIC;
}

**Comment by zuckschwerdt** [comment]
The fail code shoud be SANITY not MIC :)
We generally limit only known bad values, not implausible ones.
A sensor might work with 60+ °C? If you limit to 60.0 then (600+400) you limit the field to 10 bit, but it's 12 bit -- or are the upper bits something else? How would `< -40` work with `unsigned - 400` :)
Some sensors report humidity 0/100 on under-/overflow, is 10/99 really a protocol limit? Why is humidity 12 bit wide? I guess the upper bits are something else, right?

**Comment by some-random-kahuna** [comment]
Hi Christian,

All good questions.  

>The fail code shoud be SANITY not MIC :)
OK (with me at least)

>A sensor might work with 60+ °C? If you limit to 60.0 then (600+400) you limit the field to 10 bit, but it's 12 bit -- or are the upper bits something else? 

I picked the numbers based on the sensor environmental specifications mentioned earlier in the module.  As for how the fields are *really* defined, your guess is as good as mine.  Based on the operating specs, though, 2^10 seems more plausible than 2^12 for temperature and a single byte would do it for humidity.  Where did the field definitions for the TH3 come from?

>How would < -40 work with unsigned - 40

raw_temp is an int, so the conversion and subsequent limit check should be copacetic.

I will fully agree that this is a dirty hack and that we're missing something about the bits coming in from the TH2 sensor.  Suggestion?  More raw captures?

Aloha,

-Mark

**Comment by zuckschwerdt** [comment]
To clarify on the `< -40`, a recent compiler or linter will be smart enough to point out that `unsigned - 40` will never match that condition ;)

**Pull request merged**